### DISCUSSION TITLE
Type error: Type `ResolvedMetadata` is not assignable to type `Metadata`.

### DISCUSSION THREAD

**Discussion by filingroove** [vercel/next.js#48187] [Help]
### Summary

Confusing behaviour of parent metadata in `generateMetadata`.

Just passing parent metadata without any change will not work because of type mismatch.
```typescript
export async function generateMetadata(params: Props, parent: ResolvingMetadata): Promise<Metadata> {
  const parentMetadata = await parent;

  return {
    ...parentMetadata
  }
}
```

```shell
Type error: Type 'ResolvedMetadata' is not assignable to type 'Metadata'.
  Types of property 'robots' are incompatible.
    Type 'ResolvedRobots | null' is not assignable to type 'string | Robots | null | undefined'.
      Type 'ResolvedRobots' is not assignable to type 'string | Robots | null | undefined'.
        Type 'ResolvedRobots' is not assignable to type 'Robots'.
          Type 'ResolvedRobots' is not assignable to type '{ googleBot?: string | RobotsInfo | undefined; }'.
            Types of property 'googleBot' are incompatible.
              Type 'string | null' is not assignable to type 'string | RobotsInfo | undefined'.
                Type 'null' is not assignable to type 'string | RobotsInfo | undefined'.
```

Why is `Metadata` different from `ResolvedMetadata`? 
Should I be using `ResolveMetadata`/`Promise<ResolvedMetadata>` as  return type of `generateMetadata`?

**Comment by arindammitra06** [comment]
I am facing this too!!

**Comment by alanhamlett** [comment]
Use `Promise<Metadata | ResolvedMetadata>`.

**Comment by halu103** [comment]
import { Metadata, ResolvingMetadata } from 'next';
import { getSeo } from '@/api/seo';
interface BlogLayoutProps {
  children: React.ReactNode;
}
type Props = {
  title: string;
  description: string;
  setIndex: string;
  icon: string;
  logo: string;
};
export async function generateMetadata(
  parent: ResolvingMetadata
): Promise<Metadata | ResolvingMetadata> {
  const repo: Props = await getSeo('blog');
  const prevTitle = (await parent)?.title || 'Blog';
  const prevDes = (await parent)?.description || undefined;
  const previousImages = (await parent)?.openGraph?.images || [];
  const prevIcons = (await parent)?.icons;
  return {
    title: repo?.title ? repo.title : prevTitle,
    description: repo?.description ? repo.description : null,
    openGraph: {
      title: repo?.title ? repo.title : prevTitle,
      images: repo?.logo ? [repo.logo, ...previousImages] : [...previousImages],
      description: repo?.description ? repo.description : prevDes,
    },
    icons: repo?.icon ? repo.icon : prevIcons,
  };
}
export default async function BlogLayout({
  children,
}: BlogLayoutProps): Promise<JSX.Element> {
  return <>{children}</>;
}


Type error: Type 'OmitWithTag<ResolvingMetadata, keyof LayoutProps, "generateMetadata">' does not satisfy the constraint '{ [x: string]: never; }'.
  Property 'then' is incompatible with index signature.
    Type '<TResult1 = ResolvedMetadata, TResult2 = never>(onfulfilled?: ((value: ResolvedMetadata) => TResult1 | PromiseLike<TResult1>) | null | undefined, onrejected?: ((reason: any) => TResult2 | PromiseLike<...>) | ... 1 more ... | undefined) => Promise<...>' is not assignable to type 'never'.

  30 | // Check the arguments and return type of the generateMetadata function
  31 | if ('generateMetadata' in entry) {
> 32 |   checkFields<Diff<LayoutProps, FirstArg<MaybeField<TEntry, 'generateMetadata'>>, 'generateMetadata'>>()
     |               ^
  33 |   checkFields<Diff<ResolvingMetadata, SecondArg<MaybeField<TEntry, 'generateMetadata'>>, 'generateMetadata'>>()
  34 | }
  35 |
  
  
  what is my problem? help me !!!

**Reply by natngu18** [reply]
`export async function generateMetadata(
    {}: Props,
    parent: ResolvingMetadata
): Promise<Metadata>`

I believe parent should be second param, it fixed my error when deploying

**Comment by shehi** [comment]
Doesn't work:

```
Type error: Type '{ title: string; description: string; metadataBase: null | URL; applicationName: null | string; authors: null | Array<Author>; generator: null | string; keywords: null | Array<string>; ... 25 more ...; 'apple-touch-icon-precomposed'?: never; }' is not assignable to type 'Metadata | ResolvedMetadata'.
  Type '{ title: string; description: string; metadataBase: null | URL; applicationName: null | string; authors: null | Array<Author>; generator: null | string; keywords: null | Array<string>; ... 25 more ...; 'apple-touch-icon-precomposed'?: never; }' is not assignable to type 'ResolvedMetadata'.
    Types of property 'title' are incompatible.
      Type 'string' is not assignable to type 'AbsoluteTemplateString'.
```
